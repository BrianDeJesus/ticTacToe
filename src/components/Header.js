import React from 'react';

const Header = () => (
    <div className="App-header">
        <a href="https://github.com/BrianDeJesus/ticTacToe" className="Github-link">View on Github</a>
        <h1 className="App-title">Tic-Tac-Toe</h1>
    </div>
);

export default Header;