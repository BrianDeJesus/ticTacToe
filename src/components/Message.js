import React from 'react';
import { connect } from 'react-redux';

const Message = (props) => {
    const ending = props.gameEnding;
    let ender;
    switch(ending) {
        case 'draw': 
            ender = 'It\'s a draw!';
            break;
        case 'human':
            ender = 'You Won!';
            break;
        case 'ai':
            ender = 'You lost!';
            break;
    };
    return (
        <div>
            {
            !!props.gameEnding ? ( <p>{ender}</p> ) :
            (
                <p>It's {props.whoseTurn}'s turn</p>
            )
            }
        </div>
    );
};

const mapStateToProps = (state) => ({
    gameEnding: state.gameEnding,
    whoseTurn: state.currentTurn.turn
});

export default connect(mapStateToProps)(Message);