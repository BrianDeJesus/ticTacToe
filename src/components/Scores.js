import React from 'react';
import { connect } from 'react-redux';

export const Scores = (props) => {
    return (
        <div className="Score-container">
            <span className="Human-score"> 
                You: {props.humanScore}
            </span>
            <span className="Ai-score"> 
                AI: {props.aiScore}
            </span>
        </div>
    );
};

const mapStateToProps = (state) => ({
    humanScore: state.scores.humanScore,
    aiScore: state.scores.aiScore
});

export default connect(mapStateToProps)(Scores);