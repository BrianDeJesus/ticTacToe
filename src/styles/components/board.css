.Board {
    margin: 0 auto;
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    height: 300px;
    width: 300px;
    font-size: 3rem;
}

.Square:hover {
    background-color: rgba(163, 160, 160, 0.24);
}