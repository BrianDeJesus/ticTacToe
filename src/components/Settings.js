import React from 'react';
import { connect } from 'react-redux';
import { setFirstPlayer, getNextPlayer } from '../actions/turn';
import { renderAiChoice, clearBoard } from '../actions/board';

class Settings extends React.Component {

onPlayAsChange = (e) => {
    this.props.dispatch(clearBoard());
    if(e.target.value === 'O') {
        this.props.dispatch(setFirstPlayer('ai'));
        this.props.dispatch(renderAiChoice('X', this.props.board, 2));
        this.props.dispatch(getNextPlayer('ai'));
    }
};

 render() {
     return (
        <div>
            <span>Difficulty:</span>
            <select>
                <option value="insane">Insane</option>
                <option value="normal">Normal</option>
            </select>
            <span> Play as X's or O's? </span>
            <select onChange={this.onPlayAsChange}>
                <option value="X">X</option>
                <option value="O">O</option>
            </select>
        </div>
    );
 };
};

const mapStateToProps = (state) => ({
    currentTurn: state.currentTurn,
    board: state.board
});

export default connect(mapStateToProps)(Settings);
