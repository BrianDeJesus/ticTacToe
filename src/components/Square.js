import React from 'react';
import { connect } from 'react-redux';
import { renderClickedSpot } from '../actions/board';
import { setFirstPlayer, asyncNextPlayer } from '../actions/turn';
import computeWinningCombo from '../computations/computeWinningCombo';
import computeDraw from '../computations/computeDraw';
import { setDrawEnding, setWinner } from '../actions/gameEnding';
import { setHumanScore } from '../actions/score';

export class Square extends React.Component {

    onClickSquare = () => {
        if(this.props.currentTurn.currentPlayer === 'human' && (this.props.board[this.props.id] === '') && (computeWinningCombo(this.props.board) === "Z")) {
            this.props.dispatch(renderClickedSpot(this.props.id, this.props.currentTurn.turn));
            let winner;
            setTimeout(() => { //take a little time to get new changes
                winner = this.props.board[computeWinningCombo(this.props.board)[0]];
                if(winner !== undefined) {
                    this.props.dispatch(setWinner(this.props.currentTurn.currentPlayer));
                    this.props.dispatch(setHumanScore());
                }
            }, 90);
            if(computeDraw(this.props.board)) {
                this.props.dispatch(setDrawEnding());
            }
            this.props.dispatch(asyncNextPlayer(this.props.currentTurn.currentPlayer));
            setTimeout(() => {
                this.props.compResponse();
            }, 400); // wait 4 ms due to async next player being 3 ms
        }
    };

    render() {
        return ( 
            <div>
                <div 
                onClick={this.onClickSquare} 
                className={computeWinningCombo(this.props.board).includes(this.props.id) ? 'Winning-square' : 'Square'}
                >
                    {
                        this.props.board[this.props.id]
                    }
                </div>
            </div>
            );
        };
}

const mapStateToProps = (state) => {
    return {
        board: state.board,
        currentTurn: state.currentTurn
    };
};

export default connect(mapStateToProps)(Square);