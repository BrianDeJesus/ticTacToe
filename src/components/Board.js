import React from 'react';
import Square from './Square';
import { connect } from 'react-redux';
import { renderAiChoice } from '../actions/board';
import { setFirstPlayer, asyncNextPlayer } from '../actions/turn';
import computeWinningCombo from '../computations/computeWinningCombo';
import computeDraw from '../computations/computeDraw';
import { setDrawEnding, setWinner } from '../actions/gameEnding';
import { setAiScore } from '../actions/score';

export class Board extends React.Component {

    compResponse = () => { //AI response to human click
        if(this.props.currentTurn.currentPlayer === 'ai' && (computeWinningCombo(this.props.board) === "Z")) {
            this.props.dispatch(renderAiChoice(this.props.currentTurn.turn, this.props.board, this.props.gameDifficulty));
            const winner = this.props.board[computeWinningCombo(this.props.board)[0]];
            if(winner !== undefined) {
                this.props.dispatch(setWinner(this.props.currentTurn.currentPlayer));
                this.props.dispatch(setAiScore());
            }
            else if(computeDraw(this.props.board)) {
                this.props.dispatch(setDrawEnding());
            }
            this.props.dispatch(asyncNextPlayer(this.props.currentTurn.currentPlayer));
        }
    };
    render() {
        return  (
            <div>
                <div className="Board" >
                    {
                        this.props.board.map((square, index) => {
                            return <Square key={index} id={index} compResponse={this.compResponse} />
                        })
                    }
                </div>
            </div>
        );
    };
}

const mapStateToProps = (state) => {
    return {
        board: state.board,
        currentTurn: state.currentTurn,
        gameDifficulty: state.gameDifficulty
    };
};

export default connect(mapStateToProps)(Board);