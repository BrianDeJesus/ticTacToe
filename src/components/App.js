import React from 'react';
import Header from './Header';
import Board from './Board';
import Settings from './Settings';

const App = () => {
    return (
      <div className="App">
        <Header />
        <Settings />
        <Board />
      </div>
    );
};

export default App;
