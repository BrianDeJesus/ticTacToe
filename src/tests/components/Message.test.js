import React from 'react';
import { Message } from '../../components/Message';
import shallowRenderer from 'react-test-renderer/shallow';

test('should render Message component', () => {
    const renderer = new shallowRenderer();
    const component = renderer.render(<Message />);
    expect(component).toMatchSnapshot();
});