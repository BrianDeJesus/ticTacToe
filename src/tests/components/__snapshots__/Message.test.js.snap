// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Message component 1`] = `
<div
  className="Message"
>
  <p>
    It's 
    's turn
  </p>
</div>
`;
